### ISSUE TITLE
Location updates blocking when in battery saver mode on Android

### ISSUE THREAD

**Issue by smartyw** [organicmaps/organicmaps#2342]
I always have my Android phone in Battery Saver mode when using Organic Maps. I've been using the application for about 18 months and it has historically worked well with Battery Saver mode. I have the display set to sleep after 60 seconds and the usual behaviour is that if I want to check the map I double tap my phone's screen to wake it up and show the Organic Maps map. Location services are suspended when the screen is asleep and Battery Saver mode is enabled so I am accustomed to then waiting for around 5 - 10 seconds for a location update to be acquired at which point the map centres on my current location.

I am currently hiking a multi-day 100 mile trail and the application on this trip is behaving differently to the previous description. When I wake the screen and show Organic Maps (which is already in foreground), when a location update starts it seems to block and never complete. This does not happen every time but it does happen about 90% of the time (I have Battery Saver enabled again). I sometimes also see a dialogue which says Please Enable Location Services" but they are already enabled. Again, this does not happen every time.

![Screenshot_20220403-134058](https://user-images.githubusercontent.com/1387793/161552797-f678d4e9-3e83-469b-be10-a59c5db6459b.jpg)

To resolve the issue I have to kill the application and restart it. On restarting, my location is obtained pretty much instantaneously.

I have not looked at code but suspect a race condition. When the application comes into foreground on awakening the phone from its sleeping state, it seem to request a location update but maybe this is before Android has re-enabled the service on coming out of the Battery Saver sleep state. My theory may not be correct but it should give a good sense of the behaviour I'm experiencing on this trip. It has never been like this before and it's over six months since I last used the app on a trip so I'm thinking something has changed.

It's easy to recreate as described but make sure no other sources of location (e.g. Wi-Fi scanning) are enabled in the Android Settings Location screen otherwise you may find you cannot recreate the problem. I'm assuming it relates to GPS specifically.

My phone runs Android 10 on a OnePlus 5.

The Help screen says 2022-03.23-4-Google.

Happy to help look for a solution if this is accepted as a bug.

**Comment by smartyw** [comment]
You're welcome. And thank you for such a fast response :-)

Will do. Currently cloning the repo to my Chromebook so I can take a look at the code out of interest if nothing else.

I'll do as you ask and you should have the details tomorrow when I resume my hike at the latest. I have another trip planned for late June this time cycling so hoping to resolve this by then.

All the best,

M

**Comment by biodranik** [comment]
Thank you for the logs. Looks like the issue is caused by the enabled Power Saving mode that blocks using high precision location permission and requires the app to ask a user about explicitly granting this permission.
The error in our log: 
`GoogleFusedLocationProvider: Service is not available: com.google.android.gms.common.api.ResolvableApiException: 6: RESOLUTION_REQUIRED`

The possible fix:
```Java
@Override
  public void onFailure(@NonNull Exception e) {
    int statusCode = ((ApiException) e).getStatusCode();
    if (statusCode
        == LocationSettingsStatusCodes
        .RESOLUTION_REQUIRED) {
      // Location settings are not satisfied, but this can
      // be fixed by showing the user a dialog
      try {
        // Show the dialog by calling
        // startResolutionForResult(), and check the
        // result in onActivityResult()
        ResolvableApiException resolvable =
            (ResolvableApiException) e;
        resolvable.startResolutionForResult
            ((Activity) context,
                REQUEST_CODE);
      } catch (IntentSender.SendIntentException sendEx) {
        // Ignore the error
      }
    }
  }
});
```

**Comment by biodranik** [comment]
Nobody speaks about "always" permission. As far as I understood, the user should confirm that he allows "precise" permission instead of a "rough" one even when the battery saving mode is enabled. You expect to see a precise location on the map after turning the screen on again, right?

**Comment by smartyw** [comment]
Yes, absolutely :-) I'm just wondering if granting the permission would have any impact on the effectiveness of battery saver mode.

**Comment by rtsisyk** [comment]
Relevant discussion: https://github.com/android/location-samples/issues/193#issuecomment-482646718

I tested two Android 12 devices with Google Location Provider and Power Save mode - both works Ok. It seems that this issue only reproduces on specific devices and/or older Android versions.

> So, before Android 9.0 Pie there were 3 levels of localisation accuracy that the user can define in his own settings of the device :
> 
>    - ONLY GPS (Device Only)
>    - WIFI AND NETWORK (Battery saving)
>    - GPS, WIFI AND NETWORK (High)

@smartyw do you have "Use High-precision mode" in Location settings? If yes, could you please how this option affects the behaviour?

**Comment by smartyw** [comment]
Hi @rtsisyk ,

I'm running Android 10 and in Settings/Location have "Wi-Fi and Bluetooth scanning" with Wi-Fi Scanning enabled and Bluetooth off. In Settings/Location/Advanced I have Google Location Accuracy which is either on or off and I have it set to ON.

![Screenshot_20220411-143253](https://user-images.githubusercontent.com/1387793/162752273-1cf5fa25-7964-4f83-8433-51987a502df0.jpg)

I already tested with Wi-Fi scanning and when it was enabled and I was in a place with Wi-Fi the problem did not occur. If I disabled Wi-Fi scanning then it did occur. So I concluded it only related to the time it takes to re-enable GPS under the facade of the Google Location Services. I guess the exception in the logs illustrates clearly that an attempt to use GPS was taking place whilst still in a power saving mode. So the app is coming into foreground and the Android activity lifecycle call takes place before the power saving mode has been exited. Or something like that :-)

**Comment by rtsisyk** [comment]
We fixed `com.google.android.gms.common.api.ResolvableApiException: 6: RESOLUTION_REQUIRED` by #2342. Actually, this new dialog just advise you to enable Wi-Fi and doesn't bring any other value. It will be available in the next release. The original problem has been resolved by 44260ae3dfde048f4ad11f68c57aa1a6a5290d1c which is included since May 2022 release.

**Comment by smartyw** [comment]
Excellent, thank you. FYI I recently completed a 4,500 Km bikepacking trip around Europe and used Organic Maps for all navigation. It was invaluable! 

See [https://the-bikepacker.blogspot.com/p/europe-2022-index.html](https://the-bikepacker.blogspot.com/p/europe-2022-index.html)
